Source/Core/InputCommon/ControllerInterface/Wayland/Wayland.cpp

```diff
@@ -0,0 +1,386 @@
+    Key* key = nullptr;
+    for (xkb_layout_index_t layout = 0; layout < num_layouts && !key; layout++)
+    {
+      const xkb_level_index_t num_levels =
+          xkb_keymap_num_levels_for_key(m_xkb_keymap, keycode, layout);
+      if (num_levels == 0)
+        continue;
+
+      // Take the first level which we find a valid keysym for.
+      for (xkb_level_index_t level = 0; level < num_levels; level++)
+      {
+        const xkb_keysym_t* keysyms;
+        int num_syms =
+            xkb_keymap_key_get_syms_by_level(m_xkb_keymap, keycode, layout, level, &keysyms);
+        if (num_syms == 0)
+          continue;
+
+        // Just take the first. Should only be one in most cases anyway.
+        const xkb_keysym_t keysym = xkb_keysym_to_upper(keysyms[0]);
+
+        char keysym_name_buf[64];
+        if (xkb_keysym_get_name(keysym, keysym_name_buf, sizeof(keysym_name_buf)) <= 0)
+          continue;
+
+        const u32 bit_index = static_cast<u32>(keycode - m_min_keycode);
+        const u32 array_index = bit_index / 32;
+        const u32 mask = 1u << (bit_index % 32);
+        key = new Key(keysym_name_buf, keycode, m_state, array_index, mask);
+        AddInput(key);
+        break;
+      }
+    }
+  }
+}
+
+void KeyboardMouse::SeatCapabilities(void* data, wl_seat* seat, uint32_t capabilities)
+{
+  KeyboardMouse* kbm = static_cast<KeyboardMouse*>(data);
+  if (capabilities & WL_SEAT_CAPABILITY_KEYBOARD)
+  {
+    static const wl_keyboard_listener keyboard_listener = {
+        &KeyboardMouse::KeyboardKeymap, &KeyboardMouse::KeyboardEnter,
+        &KeyboardMouse::KeyboardLeave, &KeyboardMouse::KeyboardKey,
+        &KeyboardMouse::KeyboardModifiers};
+    kbm->m_wl_keyboard = wl_seat_get_keyboard(seat);
+    wl_keyboard_add_listener(kbm->m_wl_keyboard, &keyboard_listener, kbm);
+  }
+  if (capabilities & WL_SEAT_CAPABILITY_POINTER)
+  {
+    static const wl_pointer_listener pointer_listener = {
+        &KeyboardMouse::PointerEnter, &KeyboardMouse::PointerLeave, &KeyboardMouse::PointerMotion,
+        &KeyboardMouse::PointerButton, &KeyboardMouse::PointerAxis};
+    kbm->m_wl_pointer = wl_seat_get_pointer(seat);
+    wl_pointer_add_listener(kbm->m_wl_pointer, &pointer_listener, kbm);
+  }
+}
+
+void KeyboardMouse::GlobalRegistryHandler(void* data, wl_registry* registry, uint32_t id,
+                                          const char* interface, uint32_t version)
+{
+  KeyboardMouse* kbm = static_cast<KeyboardMouse*>(data);
+  if (std::strcmp(interface, "wl_seat") == 0)
+  {
+    static const wl_seat_listener seat_listener = {&KeyboardMouse::SeatCapabilities};
+    kbm->m_wl_seat = static_cast<wl_seat*>(wl_registry_bind(registry, id, &wl_seat_interface, 1));
+    wl_seat_add_listener(kbm->m_wl_seat, &seat_listener, kbm);
+  }
+}
+
+void KeyboardMouse::GlobalRegistryRemover(void* data, wl_registry* registry, uint32_t id)
+{
+}
+
+std::string KeyboardMouse::GetName() const
+{
+  return "Seat";
```

Comment:
Despite perhaps being accurate, "Seat" is maybe a weird name to display to our users.